### ISSUE TITLE
cid is not part of the global state

### ISSUE THREAD

**Issue by kanithin** [ipfs/ipfs-cluster#499]
#### Basic information

* [x] Version information (mark as appropiate):
  * [ ] Master
  * [ ] Release candidate for next version
  * [x] Latest stable version
  * [ ] An older version I should not be using
* [x] Type (mark as appropiate):
  * [x] Bug
  * [ ] Feature request
  * [ ] Enhancement
* [x] Operating system (mark as appropiate):
  * [x] Linux
  * [ ] macOS
  * [ ] Windows
  * [ ] Other: `which?`
* [x] Installation method (mark as appropiate):
  * [x] Binaries from dist.ipfs.io
  * [ ] Built from sources
  * [ ] Docker
  * [ ] Snap
  * [ ] Other: `which?`

#### Description
Hello,
I am trying to use the go client for IPFS Cluster http API for testing how the pinning works in the cluster.
Following is the sample code

```
cid,_ := cc.Decode(mhash)
clusterClient,_:= cl.NewClient(cfg)
err = clusterClient.Pin(cid, minReplicationFactor, maxReplicationFactor,handler.Filename)
if err != nil {
	panic(err)
}
```

When I execute this code, I always get the below error response from the server. 

```
_2018/08/13 17:48:09 http: panic serving 127.0.0.1:39636: cid is not part of the global state (404)
goroutine 21 [running]:_
```

These are the logs from the cluster

```
_17:48:09.172  INFO    cluster: IPFS cluster pinning QmTv2Tx9XQeLrvg8rs9LCCih6FrHt2mXs3LVBt23ZD7eE7 on [<peer.ID fT3SAJ> <peer.ID ZFk7GT> <peer.ID ctiUyZ> <peer.ID NdcjFP> <peer.ID XcjNQi>]: cluster.go:913
17:48:09.178 ERROR  p2p-gorpc: cid is not part of the global state client.go:245
17:48:09.178 ERROR    restapi: sending error response: 404: cid is not part of the global state restapi.go:581
17:48:09.273  INFO   ipfshttp: IPFS Pin request succeeded: QmTv2Tx9XQeLrvg8rs9LCCih6FrHt2mXs3LVBt23ZD7eE7 rpc_api.go:247_
```

Is there any reason for this "cid is not part of the global state" error?

**Comment by hsanjuan** [comment]
Hi,

I can't fully explain this one. First:

`17:48:09.172  INFO    cluster: IPFS cluster pinning QmTv2Tx9XQeLrvg8rs9LCCih6FrHt2mXs3LVBt23ZD7eE7 on [<peer.ID fT3SAJ> <peer.ID ZFk7GT> <peer.ID ctiUyZ> <peer.ID NdcjFP> <peer.ID XcjNQi>]: cluster.go:913`

says it's line 913 of `cluster.go`, but in version `0.4.0` this log message should come from 960.

Second:

`17:48:09.178 ERROR    restapi: sending error response: 404: cid is not part of the global state restapi.go:581`

This should be in line 830.

And:

`17:48:09.273  INFO   ipfshttp: IPFS Pin request succeeded: QmTv2Tx9XQeLrvg8rs9LCCih6FrHt2mXs3LVBt23ZD7eE7 rpc_api.go:247`

There the message is definitely not from rpc_api.go. It's from the `httpipfs.go` file. 

Your error can only happen when doing a PinGet operation (GET /allocations/<hash>), but the Client.Pin operation performs a `POST /pins/<hash>`.

In short, I have no idea what's going on in your setup. Can you double check? Can you send full logs running with `ipfs-cluster-service --log-level debug daemon` and setting your Client's `cfg.LogLevel` to `debug` ?

**Comment by hsanjuan** [comment]
Indeed, forget about line numbers, I see similar things happening in our logs. In any case, that error can only come from a `PinGet` and not from a `Pin` operation.

**Comment by kanithin** [comment]
Hi hsanjuan,

These are the cluster service log with logging set as debug

_08:24:53.379 DEBUG    restapi: rest api pinHandler: QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3 restapi.go:340
08:24:53.379 DEBUG    cluster: obtainAllocations: current valid: 0 allocate.go:74
08:24:53.379 DEBUG    cluster: obtainAllocations: candidates valid: 6 allocate.go:74
08:24:53.379 DEBUG    cluster: obtainAllocations: Needed: 3 allocate.go:74
08:24:53.379 DEBUG    cluster: obtainAllocations: Wanted: 5 allocate.go:74
08:24:53.379 DEBUG    cluster: obtainAllocations: allocate(): [<peer.ID fT3SAJ> <peer.ID ctiUyZ> <peer.ID ZFk7GT> <peer.ID ctdkno> <peer.ID XcjNQi> <peer.ID NdcjFP>] allocate.go:74
08:24:53.379  INFO    cluster: IPFS cluster pinning QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3 on [<peer.ID fT3SAJ> <peer.ID ctiUyZ> <peer.ID ZFk7GT> <peer.ID ctdkno> <peer.ID XcjNQi>]: cluster.go:913
08:24:53.379 DEBUG  consensus: attempt #0: committing &{Cid:{Cid:QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3 Name:NZ.jpeg Allocations:[QmfT3SAJJyPvACGTLWyckVuiav4zSwp15d2rkadPa2z37Q QmctiUyZ5fpZf1GwqrrUVJ1Mrc4ArNordKLJzbSi44ZK8Q QmZFk7GT2SLLEjsVQfjTzzpNVHnPUiHRdiRevKSxpEypXv QmctdknoK6b5qCmQC6ftwNCRWHmn539ip2QHFhz5ooTU2P QmXcjNQimbqz64xz76r4TfBAs8Fvj45GxcUuHXVTmFxUWg] ReplicationFactorMin:3 ReplicationFactorMax:5 Recursive:true} Type:1 consensus:<nil>} consensus.go:319
08:24:53.379 DEBUG  consensus: redirect try 0 consensus.go:286
08:24:53.379 DEBUG  consensus: redirecting ConsensusLogPin to leader: QmZFk7GT2SLLEjsVQfjTzzpNVHnPUiHRdiRevKSxpEypXv consensus.go:286
08:24:53.386 DEBUG    restapi: rest api pinHandler done restapi.go:340
08:24:53.387 ERROR  p2p-gorpc: cid is not part of the global state client.go:245
08:24:53.387 ERROR    restapi: sending error response: 404: cid is not part of the global state restapi.go:581
08:24:53.442 DEBUG pintracker: tracking QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3 rpc_api.go:201
08:24:53.442 DEBUG pintracker: issuing pin call for QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3 maptracker.go:65
08:24:53.442 DEBUG   ipfshttp: posting pin/ls?arg=QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3&type=recursive ipfshttp.go:759
08:24:53.443 DEBUG   ipfshttp: posting refs?arg=QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3&recursive=true ipfshttp.go:775
08:24:53.446 DEBUG   ipfshttp: Refs for QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3 sucessfully fetched rpc_api.go:247
08:24:53.446 DEBUG   ipfshttp: posting pin/add?arg=QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3&recursive=true ipfshttp.go:759
08:24:53.472  INFO   ipfshttp: IPFS Pin request succeeded: QmcrieJjKMxRLNS1ABvXi8AbsYuxuPbH32edkDWywv6jw3 rpc_api.go:247_

And I use this Pin call [https://godoc.org/github.com/ipfs/ipfs-cluster/api/rest/client#Client.Pin](https://godoc.org/github.com/ipfs/ipfs-cluster/api/rest/client#Client.Pin)

It looks like this Pin call calls PinGet internally??
But I also see the pin request succeed

**Comment by hsanjuan** [comment]
There is no codepath (that I can find) in the Pin endpoint that would make a PinGet. I'm pretty sure about this. I can't reproduce it either and we use this call a lot in our pinbot integrations without problems.

`rest api pinHandler done restapi.go:340` <- this indicates that the pin handler is done and should return 202 Accepted.

What comes below (the error message) is consistent with a call of `Client.Allocation(cid)` or GET /allocations/<cid>, but not with the Pin call above.

Is there anything special about your setup? Are you also using the `api/rest/client` from 0.4.0? what happens when using `ipfs-cluster-ctl pin add` instead?

**Comment by kanithin** [comment]
Hsanjuan, Thank you for your time. It was so silly of me that I did'nt notice that I was calling the allocations function. I removed that function call and things are working as expected. Thanks again for your time.
